### PR TITLE
Spark Operator example update with best practices

### PR DESCRIPTION
**Pull request by vara-bonthu**
### What does this PR do?

Spark Operator example updated with best practices
 - Multiple node groups (Core and Spark)
 - VPC endpoints for worker node private access 
-  Addons config updated to support large clusters e.g., coredns, fluentbit etc.
- Spark team created to support the Sample Spark jobs
- Spark Operator manifest files updated 

Closes Issue: #850

🛑 Please open an issue first to discuss any significant work and flesh out details/direction - we would hate for your time to be wasted.
Consult the [CONTRIBUTING](https://github.com/aws-ia/terraform-aws-eks-blueprints/blob/main/CONTRIBUTING.md#contributing-via-pull-requests) guide for submitting pull-requests.

### More

- [x] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [ ] Yes, I have added a new example under [examples](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples) to support my PR
- [ ] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/eks-blueprints-add-ons) repo (if applicable)
- [ ] Yes, I have updated the [docs](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/docs) for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR

**Note**: Not all the PRs required examples and docs except a new pattern or add-on added.

### For Moderators

- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

File: examples/analytics/spark-k8s-operator/spark-samples/pyspark-pi-job-yunikorn.yaml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+metadata:
+  name: catalog-sales-test1
+  namespace: spark-team-a
+spec:
+  batchScheduler: "yunikorn"
+  type: Python
+  sparkVersion: "3.1.1"
+  pythonVersion: "3"
+  mode: cluster
+  image: "public.ecr.aws/r1l5w1y9/spark-operator:3.2.1-hadoop-3.3.1-java-11-scala-2.12-python-3.8-latest"
```
Comment (askulkarni2) [inline examples/analytics/spark-k8s-operator/spark-samples/pyspark-pi-job-yunikorn.yaml:13]: What repo is this? If not an official image I recommend adding Dockerfile and relevant steps to build and push the image to ECR repo.

File: examples/analytics/spark-k8s-operator/spark-samples/tpcds-benchmark-1t.yaml [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  namespace: spark-team-a
+  labels:
+    yunikorn.apache.org/username: "vara"
+    queue: "root.spark-team-a"
+spec:
+  type: Scala
+  pythonVersion: "3"
+  sparkVersion: "3.1.2"
+  mode: cluster
+  image: ghcr.io/abin-tiger/spark-tpc-ds:3.1.2
```
Comment (askulkarni2) [inline examples/analytics/spark-k8s-operator/spark-samples/tpcds-benchmark-1t.yaml:14]: Same repo question as above.

**Review by askulkarni2 — COMMENTED** [review]
@vara-bonthu looking good, some comments.

File: examples/analytics/spark-k8s-operator/helm-values/aws-for-fluentbit-values.yaml [hunk 3]
```diff
@@ -0,0 +1,162 @@
+global:
+  ## Override the deployment namespace
+  namespaceOverride: logging
+
+image:
+  repository: amazon/aws-for-fluent-bit
```
[the remark this answers is not in the document]
Comment (vara-bonthu) [reply examples/analytics/spark-k8s-operator/helm-values/aws-for-fluentbit-values.yaml:6]: Just noticed the new repo url. Will update that

File: examples/analytics/spark-k8s-operator/helm-values/cluster-autoscaler-values.yaml [hunk 4]
```diff
@@ -0,0 +1,52 @@
+autoDiscovery:
```
[the remark this answers is not in the document]
Comment (vara-bonthu) [reply examples/analytics/spark-k8s-operator/helm-values/cluster-autoscaler-values.yaml:1]: updated

File: examples/analytics/spark-k8s-operator/helm-values/coredns-autoscaler-values.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
+nameOverride: kube-dns-autoscaler
```
[the remark this answers is not in the document]
Comment (vara-bonthu) [reply examples/analytics/spark-k8s-operator/helm-values/coredns-autoscaler-values.yaml:1]: done

File: examples/analytics/spark-k8s-operator/helm-values/metrics-server-values.yaml [hunk 6]
```diff
@@ -0,0 +1,56 @@
+# HA config for metrics-server
```
[the remark this answers is not in the document]
Comment (vara-bonthu) [reply examples/analytics/spark-k8s-operator/helm-values/metrics-server-values.yaml:1]: done

File: examples/analytics/spark-k8s-operator/helm-values/prometheus-values.yaml [hunk 7]
```diff
@@ -0,0 +1,65 @@
+server:
```
[the remark this answers is not in the document]
Comment (vara-bonthu) [reply examples/analytics/spark-k8s-operator/helm-values/prometheus-values.yaml:1]: done

File: examples/analytics/spark-k8s-operator/spark-samples/pyspark-pi-job-yunikorn.yaml (hunk 1 above)
Comment (vara-bonthu) [reply examples/analytics/spark-k8s-operator/spark-samples/pyspark-pi-job-yunikorn.yaml:14]: Added comment to reference public docker imaged used

File: examples/analytics/spark-k8s-operator/main.tf [hunk 8]
```diff
@@ -91,12 +110,104 @@ module "eks_blueprints" {
...
+        subnet_type                                                      = "private"
+        "k8s.io/cluster-autoscaler/node-template/label/arch"             = "x86"
+        "k8s.io/cluster-autoscaler/node-template/label/kubernetes.io/os" = "linux"
+        "k8s.io/cluster-autoscaler/node-template/label/noderole"         = "core"
+        "k8s.io/cluster-autoscaler/node-template/label/node-lifecycle"   = "on-demand"
+        "k8s.io/cluster-autoscaler/experiments"                          = "owned"
+        "k8s.io/cluster-autoscaler/enabled"                              = "true"
+      }
+    },
+    #---------------------------------------
```
[the remark this answers is not in the document]
Comment (vara-bonthu) [reply examples/analytics/spark-k8s-operator/main.tf:153]: done

File: examples/analytics/spark-k8s-operator/spark-samples/tpcds-benchmark-1t.yaml (hunk 2 above)
Comment (vara-bonthu) [reply examples/analytics/spark-k8s-operator/spark-samples/tpcds-benchmark-1t.yaml]: updated with aws one

File: examples/analytics/spark-k8s-operator/main.tf [hunk 9]
```diff
@@ -261,8 +514,8 @@ module "vpc" {
   cidr = local.vpc_cidr
 
   azs             = local.azs
-  public_subnets  = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 8, k)]
-  private_subnets = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 8, k + 10)]
+  public_subnets  = var.public_subnets  # Two Subnets. 4094 IPs per Subnet
+  private_subnets = var.private_subnets # Three Subnets. 16382 IPs per Subnet
```
[the remark this answers is not in the document]
Comment (vara-bonthu) [reply examples/analytics/spark-k8s-operator/main.tf:518]: make sense. removed vpc_cidr from locals now.

File: examples/analytics/spark-k8s-operator/main.tf [hunk 10]
```diff
@@ -286,6 +539,80 @@ module "vpc" {
...
+    s3 = {
+      service         = "s3"
+      service_type    = "Gateway"
+      route_table_ids = module.vpc.private_route_table_ids
+      tags = {
+        Name = "${local.name}-s3"
+      }
+    }
+    },
+    { for service in toset(["autoscaling", "ecr.api", "ecr.dkr", "ec2", "ec2messages", "elasticloadbalancing", "sts", "kms", "logs", "ssm", "ssmmessages"]) :
```
[the remark this answers is not in the document]
Comment (vara-bonthu) [reply examples/analytics/spark-k8s-operator/main.tf]: updated

File: examples/analytics/spark-k8s-operator/main.tf [hunk 11]
```diff
@@ -91,12 +110,104 @@ module "eks_blueprints" {
...
+      additional_tags = {
+        Name                                                             = "core-node-grp"
+        subnet_type                                                      = "private"
+        "k8s.io/cluster-autoscaler/node-template/label/arch"             = "x86"
+        "k8s.io/cluster-autoscaler/node-template/label/kubernetes.io/os" = "linux"
+        "k8s.io/cluster-autoscaler/node-template/label/noderole"         = "core"
+        "k8s.io/cluster-autoscaler/node-template/label/node-lifecycle"   = "on-demand"
+        "k8s.io/cluster-autoscaler/experiments"                          = "owned"
+        "k8s.io/cluster-autoscaler/enabled"                              = "true"
+      }
```
[the remark this answers is not in the document]
Comment (askulkarni2) [reply examples/analytics/spark-k8s-operator/main.tf:150]: @vara-bonthu just this pending.

**Review by askulkarni2 — COMMENTED** [review]
Just one minor unresolved comment, otherwise LGTM!

File: examples/analytics/spark-k8s-operator/main.tf (hunk 11 above)
[the remark this answers is not in the document]
Comment (vara-bonthu) [reply examples/analytics/spark-k8s-operator/main.tf:150]: Added link to the doc related to ASG tags.

**Review by bryantbiggs — APPROVED** [review]

**Pull request merged**